File: packages/core/src/tools/ls.ts
import { existsSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";
import { tool } from "ai";
import { z } from "zod";
import { resolveToCwd } from "./utils/path-utils";
import { DEFAULT_MAX_BYTES, formatSize, truncateHead } from "./utils/truncate";

const DEFAULT_LIMIT = 500;

export interface LsToolContext {
  cwd: string;
  abortSignal?: AbortSignal;
}

export function createLsTool(ctx: LsToolContext) {
  return tool({
    description: `List directory contents. Returns entries sorted alphabetically, with '/' suffix for directories. Includes dotfiles. Output is truncated to ${DEFAULT_LIMIT} entries or ${DEFAULT_MAX_BYTES / 1024}KB (whichever is hit first).`,
    inputSchema: z.object({
      path: z.string().optional().describe("Directory to list (default: current directory)"),
      limit: z.number().int().positive().optional().describe("Maximum number of entries to return (default: 500)"),
    }),
    execute: async ({ path, limit }, options) => {
      const signal = options?.abortSignal ?? ctx.abortSignal;
      if (signal?.aborted) throw new Error("Operation aborted");

      const dirPath = resolveToCwd(path || ".", ctx.cwd);
      const effectiveLimit = limit ?? DEFAULT_LIMIT;

      if (!existsSync(dirPath)) throw new Error(`Path not found: ${dirPath}`);
      const st = statSync(dirPath);
      if (!st.isDirectory()) throw new Error(`Not a directory: ${dirPath}`);

      let entries: string[];
      try {
        entries = readdirSync(dirPath);
      } catch (e) {
        throw new Error(`Cannot read directory: ${e instanceof Error ? e.message : e}`);
      }
      entries.sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));

      const results: string[] = [];
      let entryLimitReached = false;
      for (const entry of entries) {
        if (signal?.aborted) throw new Error("Operation aborted");
        if (results.length >= effectiveLimit) {
          entryLimitReached = true;
          break;
        }
        const fullPath = join(dirPath, entry);
        let suffix = "";
        try {
          const entryStat = statSync(fullPath);
          if (entryStat.isDirectory()) suffix = "/";
        } catch {
          continue;
        }
        results.push(entry + suffix);
      }

      if (results.length === 0) return "(empty directory)";

      const rawOutput = results.join("\n");
      const truncation = truncateHead(rawOutput, { maxLines: Number.MAX_SAFE_INTEGER });
      let output = truncation.content;
      const notices: string[] = [];
      if (entryLimitReached) notices.push(`${effectiveLimit} entries limit reached. Use limit=${effectiveLimit * 2} for more`);
      if (truncation.truncated) notices.push(`${formatSize(DEFAULT_MAX_BYTES)} limit reached`);
      if (notices.length > 0) output += `\n\n[${notices.join(". ")}]`;
      return output;
    },
  });
}

File: internal/modules/core/tools/ls.go
package tools

import (
	"context"
	"os"
	"sort"
	"strings"

	"github.com/notshekhar/pi/internal/modules/ai"
)

type lsArgs struct {
	Path  string `json:"path,omitempty" jsonschema:"description=Directory to list; defaults to the working directory"`
	Limit *int   `json:"limit,omitempty" jsonschema:"description=Maximum entries (default 500)"`
}

// Ls returns the ls tool.
func Ls(t *Context) ai.Tool {
	return ai.NewTool("ls",
		"List directory contents, sorted alphabetically. Directories end with /. Includes dotfiles.",
		func(ctx context.Context, a lsArgs) (ai.ToolResult, error) {
			if aborted(ctx) {
				return ai.ToolError("aborted"), nil
			}
			path := Resolve(a.Path, t.CWD)
			info, err := os.Stat(path)
			if err != nil {
				return ai.ToolErrorf("path not found: %s", a.Path), nil
			}
			if !info.IsDir() {
				return ai.ToolErrorf("not a directory: %s", a.Path), nil
			}
			entries, err := os.ReadDir(path)
			if err != nil {
				return ai.ToolErrorf("cannot read directory: %v", err), nil
			}
			sort.Slice(entries, func(i, j int) bool {
				return strings.ToLower(entries[i].Name()) < strings.ToLower(entries[j].Name())
			})
			limit := 500
			if a.Limit != nil && *a.Limit > 0 {
				limit = *a.Limit
			}
			var b strings.Builder
			n := 0
			for _, e := range entries {
				if n >= limit {
					b.WriteString("\n… more entries")
					break
				}
				name := e.Name()
				if e.IsDir() {
					name += "/"
				}
				if n > 0 {
					b.WriteByte('\n')
				}
				b.WriteString(name)
				n++
			}
			if b.Len() == 0 {
				return ai.ToolText("(empty)"), nil
			}
			return ai.ToolText(b.String()), nil
		})
}
